src/Particular.HealthMonitoring.Uptime/HeartbeatFailureDetector.cs

```diff
@@ -0,0 +1,62 @@
+namespace Particular.HealthMonitoring.Uptime
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using ServiceBus.Management.Infrastructure.Settings;
+    using ServiceControl.Infrastructure;
+    using ServiceControl.Infrastructure.DomainEvents;
+
+    class HeartbeatFailureDetector : IStartable
+    {
+        EndpointInstanceMonitoring monitoring;
+        TimeSpan gracePeriod;
+        Timer timer;
+
+        public HeartbeatFailureDetector(EndpointInstanceMonitoring monitoring)
+        {
+            gracePeriod = GetHeartbeatGracePeriod();
+            this.monitoring = monitoring;
+        }
+
+        static TimeSpan GetHeartbeatGracePeriod()
+        {
+            try
+            {
+                return TimeSpan.Parse(SettingsReader<string>.Read("HeartbeatGracePeriod", "00:00:40"));
+            }
+            catch (Exception ex)
+            {
+                //logger.Error($"HeartbeatGracePeriod settings invalid - {ex}. Defaulting HeartbeatGracePeriod to '00:00:40'");
```

Comment:
Why is the log statement commented out? Still seems to be valid.

Reply:
I haven't yet figured out what to do with the logging abstraction. I would like to avoid adding a reference to `NServiceBus` in this package. On the other hand I don't like the idea of having an abstraction on top of an abstraction either.

Reply:
What happens if we just rely on native NLog? Or switch to [Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging)?

What is the danger of relying on NSB in this package? Portability between versions?

Reply:
@mikeminutillo yeah, I am worried about portability and API surface. In my mind the components should only depend on the host/engine APIs. If they need messaging (like a backplane) it needs to be an explicit API in the host.

Since in SC we use NLog I am fine with using NLog directly in components. On the other hand I guess using the Microsoft abstraction would be more future-proof...